### PR TITLE
Resources in Spin

### PR DESCRIPTION
**Pull request by rylev**
Part of #1841 

This PR moves sqlite and key-value storage to use component model resources. This is a significant breaking change to the Spin World but the user facing changes are more subtle:
* sqlite use to have `ValueParam` and `ValueResult` for values that are passed into function vs being returned back. We now only have `Value` (a direct translation of the wit) that gets passed to functions by reference and returned by ownership). So while one would have previously done `conn.execute("select * from table where foo = ?", &[ValueParam::Blob(&[1,2,3]))`, you now do `conn.execute("select * from table where foo = ?", &[Value::Blob(vec![1,2,3]))` (notice the move from a slice to an owned vector. 
* We are using the implementations more directly and thus are a bit more at the mercy of the quality of codegen from wit-bindgen. For example, `key_value::Store::get` used to take `key: impl AsRef<str>` allow for anything as long as it could be referenced as a str, but now it only takes `key: &str` allowing only for string slices (and anything dereferenced to a string slice). We could go back to wrapping these types, but I do believe one goal we've had is move to thinner and thinner SDKs so perhaps this is something we could improve on the wit-bindgen side. 

This PR only moves the Rust SDK over - we'll want to move all other SDKs over to wit-bindgen 12 as well. 

This also relies on https://github.com/fermyon/spin-componentize/pull/24

### CODE REVIEW

File: sdk/rust/src/key_value.rs [hunk 1]
```diff
@@ -5,91 +5,31 @@
...
-    /// If `name` is empty, open the default store.  Other stores may also be available depending on the app
-    /// configuration.
-    pub fn open(name: impl AsRef<str>) -> Result<Self, Error> {
-        key_value::open(name.as_ref()).map(Self)
-    }
-
-    /// Open the default store.
-    ///
-    /// This is equivalent to `Store::open("default")`.
-    pub fn open_default() -> Result<Self, Error> {
```
Comment (itowlson) [inline sdk/rust/src/key_value.rs:32]: Could we bring this back please?

File: sdk/rust/src/llm.rs [hunk 2]
```diff
@@ -7,19 +7,21 @@ pub use crate::wit::fermyon::spin::llm::{
...
+#[derive(Debug, Clone, Copy)]
 pub enum InferencingModel<'a> {
     Llama2Chat,
     CodellamaInstruct,
     Other(&'a str),
 }
 
 impl<'a> InferencingModel<'a> {
-    fn as_str(&self) -> llm::InferencingModel<'a> {
+    fn to_string(&self) -> llm::InferencingModel {
```
Comment (itowlson) [inline sdk/rust/src/llm.rs:18]: Not sure why this change is needed?

File: sdk/rust/macro/wit/reactor.wit [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package fermyon:spin
+
+world exports {
```
Comment (itowlson) [inline sdk/rust/macro/wit/reactor.wit:3]: Nothing actually exports both inbound-http and inbound-redis does it?  I recall something about this causing problems in custom triggers where the SDK wouldn't work unless you faked up an export... is this an appropriate time to address that?  (Or right time but wrong PR? _grin_)

File: crates/trigger/src/loader.rs [hunk 4]
```diff
@@ -56,6 +56,7 @@ impl Loader for TriggerLoader {
             )
         })?;
         let component = spin_componentize::componentize_if_necessary(&bytes)?;
+        std::fs::write("COMPONENT.wasm", &component).unwrap();
```
Comment (itowlson) [inline crates/trigger/src/loader.rs:59]: Left-over diagnostic code?

File: sdk/rust/macro/src/lib.rs [hunk 5]
```diff
@@ -6,17 +6,28 @@ use quote::quote;
...
+    const HTTP_COMPONENT_WIT_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/wit");
 
     quote!(
         #func
 
         mod __spin_http {
+            #![allow(missing_docs)]
+            ::spin_sdk::wit_bindgen::generate!({
+                world: "exports",
+                path: #HTTP_COMPONENT_WIT_PATH,
```
Comment (dicej) [inline sdk/rust/macro/src/lib.rs:18]: I don't think this will work in general, given that the wit directory may not be present on the machine building the application code that is invoking the macro, right?

Comment (dicej) [reply sdk/rust/macro/src/lib.rs:18]: TIL this should work since the wit directory will be packaged as part of the crate.

File: sdk/rust/src/llm.rs (hunk 2 above)
Comment (rylev) [reply sdk/rust/src/llm.rs]: It's not need but the type changed and `as_str` seemed wrong given it was returning a String and not a `str`. I've since changes this to a `Display` impl.

File: sdk/rust/src/key_value.rs (hunk 1 above)
Comment (rylev) [reply sdk/rust/src/key_value.rs:32]: Whoops!

File: sdk/rust/macro/wit/reactor.wit (hunk 3 above)
Comment (rylev) [reply sdk/rust/macro/wit/reactor.wit]: I was able to fix this 🎉

File: crates/key-value/src/lib.rs [hunk 6]
```diff
@@ -69,24 +65,33 @@ impl Default for KeyValueDispatch {
...
-    async fn get(&mut self, store: StoreHandle, key: String) -> Result<Result<Vec<u8>, Error>> {
+    async fn get(
+        &mut self,
+        store: Resource<key_value::Store>,
+        key: String,
+    ) -> Result<Result<Vec<u8>, Error>> {
         Ok(async {
             self.stores
-                .get(store)
+                .get(store.rep())
```
[the remark this answers is not in the document]
Comment (dicej) [reply crates/key-value/src/lib.rs:94]: Though I suppose that would make sharing code between the new and legacy versions more difficult.

**Review by dicej — COMMENTED** [review]
Where are the crates/trigger/wasi and crates/trigger/wit files used?  Sorry if it's obvious and I missed it.

File: crates/key-value-sqlite/src/lib.rs [hunk 7]
```diff
@@ -162,7 +163,7 @@ mod test {
         );
 
         assert!(matches!(
-            kv.exists(42, "bar".to_owned()).await?,
+            kv.exists(Resource::new_own(42), "bar".to_owned()).await?,
```
[the remark this answers is not in the document]
Comment (rylev) [reply crates/key-value-sqlite/src/lib.rs:166]: I was planning on doing this in a follow up with other changes to the wit world. I'm nervous about how big this PR has already become.

File: crates/key-value/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rylev) [reply crates/key-value/src/lib.rs:94]: I'd like to see if I can remove this in a follow up PR.

**Review by dicej — APPROVED** [review]
LGTM once we get the Wasmtime deps switched over to a snapshot of the `release-14.0.0` branch.

**Comment by rylev** [comment]
> Where are the crates/trigger/wasi and crates/trigger/wit files used? Sorry if it's obvious and I missed it.

Whoops! Those were from the component composition approach which was removed in favor of the multiple component implementation approach. I've removed those files.
